pippy/PipelineSchedule.py

```diff
@@ -179,19 +179,20 @@ def get_fwd_send_ops(self) -> List[dist.P2POp]:
             for fwd_output in self.fwd_outputs
         ]
 
-    def forward(self, args: List[torch.tensor]) -> torch.tensor:
-        logger.info(f"[{self.rank} FORWARD {self.stage_id}")
+    def forward(self, args: List[torch.tensor]) -> List[Any]:
         if self.is_first_stage:
             self.fwd_inputs = args
+        logger.info(
+            f"[{self.rank} FORWARD {self.stage_id} {[fwd_input.shape for fwd_input in self.fwd_inputs]}"
+        )
 
         # this is needed when we access the gradients for this in backward()
-        if not self.is_first_stage:
-            for tensor in self.fwd_inputs:
-                tensor.requires_grad = True
-                tensor.retain_grad()
+        for tensor in self.fwd_inputs:
+            tensor.requires_grad = True
+            tensor.retain_grad()
 
         # perform forward pass on module
-        self.fwd_outputs = self.module(*self.fwd_inputs)
+        self.fwd_outputs = [self.module(*self.fwd_inputs)]
```

Comment:
wondering why the addition of the []

(is this gonna fix one case and break another? its risky if we do not have a design written down. maybe add comment somewhere that describes our assumptions about saving and later using fwd_outputs)

Reply:
i missed this when I added the support for multiple arguments, so i was fixing it then. Each input microbatch is a list of tensors [arg1, arg2, ..], likewise, each pipeline stage output also needs to be a list of tensors [output1, output2, ...]

Reply:
let me write a comment for this

Reply:
Thanks, I fixed this in another way thats a bit more robust rather than adding the [], this will throw an error if the model forward() output is something not supported.